Dalamud/Game/Gui/Nameplates/NameplateGui.cs

```diff
@@ -0,0 +1,211 @@
+internal class NameplateGui : IInternalDisposableService, INameplatesGui
+{
+    [ServiceManager.ServiceDependency]
+    private readonly GameGui gameGui = Service<GameGui>.Get();
+
+    [ServiceManager.ServiceDependency]
+    private readonly ObjectTable objectTable = Service<ObjectTable>.Get();
+
+    [ServiceManager.ServiceDependency]
+    private readonly AddonLifecycle addonLifecycle = Service<AddonLifecycle>.Get();
+
+    [Signature("E8 ?? ?? ?? ?? E9 ?? ?? ?? ?? 48 8B 5C 24 ?? 45 38 BE", DetourName = nameof(HandleSetPlayerNameplateDetour))]
+    private readonly Hook<SetPlayerNameplateDetourDelegate> setPlayerNameplateDetourHook;
+    private nint namePlatePtr;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="NameplateGui"/> class.
+    /// </summary>
+    [ServiceManager.ServiceConstructor]
+    private NameplateGui(TargetSigScanner sigScanner)
+    {
+        // Pointers
+        this.addonLifecycle.RegisterListener(new AddonLifecycleEventListener(AddonEvent.PostSetup, "NamePlate", this.AddonLifecycle_Event));
+        this.addonLifecycle.RegisterListener(new AddonLifecycleEventListener(AddonEvent.PreFinalize, "NamePlate", this.AddonLifecycle_Event));
+
+        // Hooks
+        SignatureHelper.Initialize(this);
+        this.setPlayerNameplateDetourHook.Enable();
+    }
+
+    private unsafe delegate nint SetPlayerNameplateDetourDelegate(nint playerNameplateObjectPtr, bool isTitleAboveName, bool isTitleVisible, nint titlePtr, nint namePtr, nint freeCompanyPtr, nint prefix, int iconId);
+
+    /// <inheritdoc/>
+    public event INameplatesGui.OnNameplateUpdateDelegate OnNameplateUpdate;
+
+    /// <inheritdoc/>
+    public void DisposeService()
+    {
+        this.setPlayerNameplateDetourHook.Dispose();
+    }
+
+    /// <inheritdoc/>
+    public T? GetNameplateGameObject<T>(INameplateObject nameplateObject) where T : GameObject
+    {
+        return this.GetNameplateGameObject<T>(nameplateObject.Pointer);
+    }
+
+    /// <inheritdoc/>
+    public T? GetNameplateGameObject<T>(nint nameplateObjectPtr) where T : GameObject
+    {
+        // Get the nameplate object array
+        var nameplateObjectArrayPtrPtr = this.namePlatePtr + Marshal.OffsetOf(typeof(AddonNamePlate), nameof(AddonNamePlate.NamePlateObjectArray)).ToInt32();
+        var nameplateObjectArrayPtr = Marshal.ReadIntPtr(nameplateObjectArrayPtrPtr);
+
+        if (nameplateObjectArrayPtr == nint.Zero)
+            return null;
+
+        // Determine the index of the nameplate object within the nameplate object array
+        var namePlateObjectSize = Marshal.SizeOf(typeof(AddonNamePlate.NamePlateObject));
+        var namePlateObjectPtr0 = nameplateObjectArrayPtr + namePlateObjectSize * 0;
+        var namePlateIndex = (nameplateObjectPtr.ToInt64() - namePlateObjectPtr0.ToInt64()) / namePlateObjectSize;
+        
+        if (namePlateIndex < 0 || namePlateIndex >= AddonNamePlate.NumNamePlateObjects)
+            return null;
```

Comment:
I think it might be a good idea to provide index lookup as a separate function independent from GameObject lookup. If a plugin is doing its own modifications to plates it likely needs to know the index so it can look up the UI Node or find certain data in whatever index it uses internally.

Reply:
Maybe also makes sense, your right. I added a property to `INameplateObject`. Also splitted `NameplateGui.GetNameplateGameObject<T>` into two methods.